pkg/csi/service/wcpguest/controller.go

```diff
@@ -452,6 +467,91 @@ func (c *controller) ControllerUnpublishVolume(ctx context.Context, req *csi.Con
+			if err != nil {
+				msg := fmt.Sprintf("failed to create cnsFileAccessConfigWatcher. Error: %+v", err)
+				log.Error(msg)
+				return nil, status.Errorf(codes.Internal, msg)
+			}
+			// Attacher timeout, default is set to 4 minutes
+			timeoutSeconds := int64(getAttacherTimeoutInMin(ctx) * 60)
+			cnsFileAccessConfigInstanceName := virtualMachine.Name + "-" + req.VolumeId
+			watchCnsFileAccessConfig, err := cnsFileAccessConfigWatcher.Watch(metav1.ListOptions{
+				FieldSelector:  fields.SelectorFromSet(fields.Set{"metadata.name": cnsFileAccessConfigInstanceName}).String(),
+				TimeoutSeconds: &timeoutSeconds,
+			})
+			if err != nil {
+				msg := fmt.Sprintf("failed to watch cnsFileAccessConfig instance %q with Error: %v", cnsFileAccessConfigInstanceName, err)
+				log.Error(msg)
+				return nil, status.Errorf(codes.Internal, msg)
+			}
+			if err := c.cnsOperatorClient.Delete(ctx, &cnsfileaccessconfigv1alpha1.CnsFileAccessConfig{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      cnsFileAccessConfigInstanceName,
+					Namespace: c.supervisorNamespace,
+				},
+			}); err != nil {
+				// Delete() on the CnsFileAccessConfig instance failed with different error
+				msg := fmt.Sprintf("failed to delete CnsFileAccessConfig instance: %q. Error: %+v", cnsFileAccessConfigInstanceName, err)
+				log.Error(msg)
+				return nil, status.Errorf(codes.Internal, msg)
+			}
+
+			defer watchCnsFileAccessConfig.Stop()
+			var cnsfileaccessconfigInstanceErr string
+			timeout := time.Now().Add(time.Duration(timeoutSeconds) * time.Second)
+			// Watch all update events made on CnsFileAccessConfig instance until Deleted event or a timeout occurs on the cnsfileaccessconfig instance
+			for {
+				log.Debugf("waiting for update on cnsfileaccessconfigs: %q", cnsFileAccessConfigInstanceName)
+				event := <-watchCnsFileAccessConfig.ResultChan()
+				cnsfileaccessconfig, ok := event.Object.(*cnsfileaccessconfigv1alpha1.CnsFileAccessConfig)
+				if !ok {
+					msg := fmt.Sprintf("Watch on cnsfileaccessconfig instance %q timed out", cnsFileAccessConfigInstanceName)
+					log.Error(msg)
+					return nil, status.Errorf(codes.Internal, msg)
+				}
+				if event.Type == "DELETED" {
+					log.Infof("ControllerUnpublishVolume: Volume detached successfully %q", req.VolumeId)
+					return &csi.ControllerUnpublishVolumeResponse{}, nil
+				}
+				cnsfileaccessconfigInstanceErr = cnsfileaccessconfig.Status.Error
+				if time.Now().After(timeout) {
+					msg := fmt.Sprintf("Time out while waiting for DELETED event on cnsfileaccessconfig instance %q for volume id %q. Error = %+v", cnsFileAccessConfigInstanceName, req.VolumeId, cnsfileaccessconfigInstanceErr)
+					log.Error(msg)
+					return nil, status.Errorf(codes.Internal, msg)
+				}
+			}
+		}
+		// Feature is disabled on the cluster
+		return nil, status.Error(codes.InvalidArgument, "File volume not supported.")
+	}
+	// Block volume support
```

Comment:
Can you refactor the code to have separate methods that handle detach for block and file volumes?

Reply:
Done